### PR TITLE
fix: required action revert logic

### LINKED ISSUE #6250: `required` action in survey logic does not work properly
### Issue Summary

If a question later in the survey is made required based on a logic action, and then the response is changed to something that no longer satisfies the logic condition, the required question does not become optional again as expected.

### Expected Behavior

Changing the response should revert the logic actions, making the question optional if the condition is no longer met.

### Other information (incl. screenshots, Formbricks version, steps to reproduce,...)

Steps to reproduce:
1. Create a link survey.
2. Add a hidden field called `answer`.
3. Choose a free text question as the first question and add an optional second question of any type.
4. In the first question, create a logic rule:
  - Condition: this question’s response equals `answer` (hidden field).
5. In the “then” block, select the `require answer` action and set it for the second question.
6. While filling the survey, pre-fill the hidden field `answer` with a value, e.g., SURVEY_LINK?answer=test.
7. Answer the first question as `test` so the condition is true.
8. Notice the second question becomes required.
9. Go back and change the first question’s response to something that does **not** satisfy the condition.
10. The second question is still required, even though it should be optional again.

### Your Environment

None

### PR DESCRIPTION
**Pull request by gupta-piyush19**
## What does this PR do?
fixes required action revert logic

Fixes #6250

## Summary by CodeRabbit

* **Tests**
  * Expanded test coverage for survey logic, navigation, revert functionality, variable management, edge cases, recaptcha handling, UI elements, and error scenarios to ensure robust and correct behavior.
* **New Features**
  * Added automatic tracking and reverting of required question states when navigating back in a survey, improving dynamic survey logic and user experience.

### CODE REVIEW

File: packages/surveys/src/components/general/survey.tsx [hunk 1]
```diff
@@ -132,9 +132,23 @@ export function Survey({
...
+  const originalQuestionRequiredStates = useRef<Record<string, boolean>>({});
+
+  // state to keep track of the questions that were made required by each specific question's logic
+  const questionRequiredByMap = useRef<Record<string, string[]>>({});
+
   // Update localSurvey when the survey prop changes (it changes in case of survey editor)
   useEffect(() => {
     setlocalSurvey(survey);
+    // Update original required states when survey changes
+    originalQuestionRequiredStates.current = survey.questions.reduce<Record<string, boolean>>(
```
Comment (pandeymangg) [inline packages/surveys/src/components/general/survey.tsx]: I think changing this here in the useEffect now kind of defeats the purpose of defining this as a ref with `useRef`. I think we can just define this as a value returned from `useMemo` in which the deps array also contains the `survey` object 🙏

**Review by pandeymangg — CHANGES_REQUESTED** [review]
looks good and works well!
I just left a small comment, pls take a look 🙏

File: packages/surveys/src/components/general/survey.tsx [hunk 2]
```diff
@@ -337,6 +351,28 @@ export function Survey({
...
+            };
+          }
+          return question;
+        }),
+      }));
+
+      // remove the question from the map
+      delete questionRequiredByMap.current[questionId];
+    }
+  };
```
[the remark this answers is not in the document]
Comment (gupta-piyush19) [reply packages/surveys/src/components/general/survey.tsx]: I think it's fine, any logger would pollute with a lot of the logs here.
